lib/spack/spack/installer.py

```diff
@@ -2019,213 +2112,238 @@ def install(self) -> None:
         fail_fast_err = "Terminating after first install failure"
         single_requested_spec = len(self.build_requests) == 1
         failed_build_requests = []
-
         install_status = InstallStatus(len(self.build_pq))
+        active_tasks: List[Task] = []
 
         # Only enable the terminal status line when we're in a tty without debug info
         # enabled, so that the output does not get cluttered.
         term_status = TermStatusLine(
             enabled=sys.stdout.isatty() and tty.msg_enabled() and not tty.is_debug()
         )
+        # while a task is ready or tasks are running
+        while self._peek_ready_task() or active_tasks:
+            # while there's space for more active tasks to start
+            while len(active_tasks) < self.max_active_tasks:
+                task = self._pop_ready_task()
+                if not task:
+                    # no ready tasks ...
+                    break
```

Comment:
Can you make the main loops a bit easier to read by pulling out their bodies into nested functions?  Basically I think this will be more readable (b/c it separates the start and complete logic from the start and finish loops):

```python
        # these are nested inside install() so they can access all the local vars
        def start_task(task) -> Task:
            task = self._pop_ready_task()
            if not task:
                return False
            # ... everything in the body of the start loop below ...
            return True

        def complete_task(task):
            # ... everything in the body of the finish loop below ...
            
        # while a task is ready or tasks are running
        while self._peek_ready_task() or active_tasks:
            # while there's space for more active tasks to start
            while len(active_tasks) < self.max_active_tasks:
                start_task() or break

            time.sleep(0.1)

            # Check if any tasks have completed
            done = [task for task in active_tasks if task.poll()]

            # Iterate through the done tasks and finish them
            for task in done:
                complete_task(task)
```

This keeps most of the logic at the same level of indentation and likely cleans up the diff too.  It also lets us think about the starting and finishing of tasks without all the exception handling and stuff in the middle.